### PR TITLE
Retry getting the E2E tests Pod logs when the stream ends

### PR DESCRIPTION
**Pull request by sebgl**
In some environments (OCP mostly), we don't manage to get a full 6 hours
long Pod log stream with no interruptions. More generally, network isn't
reliable so this could happen in any environments.

Let's retry the log stream apiserver request when that happens.

The side-effect is that we now get duplicate log entries. In CI we have a
process that parses the logs to turn them into formatted test results.
In order to not make that process fail, we ask Kubernetes to
provide a timestamp with each log, so we can keep track of the last timestamp.
On a retry, we can therefore ignore logs whose timestamp is lower than
the last one we found.
This isn't a perfect system: if the stream drops in between several logs
that share the same timestamp, the second half will be lost. In our case,
for test results, I think that's unlikely to happen.

Fixes https://github.com/elastic/cloud-on-k8s/issues/3386.

### CODE REVIEW

**Comment by sebgl** [comment]
This is rather hard to test in order to ensure it fixes https://github.com/elastic/cloud-on-k8s/issues/3386...
What I did is a manual test where I spin up a Pod that logs numbers to stdout, invoke the `monitorTestJob` function for that Pod, and simulate breaks in the stream buffer once in a while: https://github.com/sebgl/cloud-on-k8s/compare/retry-e2e-logstream...sebgl:retry-e2e-logstream-manual-testing?expand=1. I then manually check that we don't process logs twice, and that the missing logs that share the same (nanosecond) timestamp are the expected ones.

File: test/e2e/cmd/run/run.go [hunk 1]
```diff
@@ -486,15 +481,84 @@ func (h *helper) streamTestJobOutput(streamStatus chan<- error, client *kubernet
...
+
+				// don't re-write logs that have been already written in a previous log stream attempt
+				if !pastPreviousLogStream && !timestamp.After(lastTimestamp) {
+					continue
+				}
+
+				// new log to process
+				pastPreviousLogStream = true
+				lastTimestamp = timestamp
+				if _, err := writer.Write([]byte(logLine + "\n")); err != nil {
```
Comment (pebrc) [inline test/e2e/cmd/run/run.go:524]: Did we use to print the timestamp as well, or am I misremembering?

Comment (sebgl) [reply test/e2e/cmd/run/run.go:524]: We did not, I'm adding it in https://github.com/elastic/cloud-on-k8s/pull/3389/files#diff-b48621daf1e1c944a38d2c599b1d11c4R543.

**Comment by sebgl** [comment]
@pebrc I did some manual tests but not a full E2E test run. I suggest we merge, trigger a nightly build and see what happens! 🤞

File: test/e2e/cmd/run/run.go (hunk 1 above)
Comment (pebrc) [reply test/e2e/cmd/run/run.go:524]: Makes sense. I missed that.

File: test/e2e/cmd/run/run.go [hunk 2]
```diff
@@ -486,15 +481,84 @@ func (h *helper) streamTestJobOutput(streamStatus chan<- error, client *kubernet
...
-	var buffer [logBufferSize]byte
-	if _, err := io.CopyBuffer(writer, stream, buffer[:]); err != nil && err != io.EOF {
-		streamStatus <- err
-		return
+
+	// The log stream may end abruptly in some environments where the network isn't reliable.
+	// Let's retry when that happens. To avoid duplicate log entries, keep track of the last
+	// Kubernetes log timestamp: we don't want to reprocess entries with a lower timestamp.
+	// If the stream drops in between several logs that share the same timestamp, the second half will be lost.
+	lastTimestamp := time.Time{}
```
Comment (charith-elastic) [inline test/e2e/cmd/run/run.go:490]:
```suggestion
	lastTimestamp := time.Now()
```

**Comment by charith-elastic** [comment]
Oops, submitted the review too late :disappointed:

**Comment by sebgl** [comment]
@charith-elastic thanks for the review! I merged a bit too soon :) I suggest we wait and see if this PR fixed https://github.com/elastic/cloud-on-k8s/issues/3407, and then decide whether we want to improve/optimize this code or not?

**Comment by charith-elastic** [comment]
:+1:  If the fix works we can optimize it in a different PR.

**Pull request merged**